man/distcc.1
.TH distcc 1 "9 June 2008"
.SH "NAME"
distcc \- distributed C/C++/ObjC compiler with distcc-pump extensions
.SH "SYNOPSIS"
.B distcc
.I <compiler> [COMPILER OPTIONS]
.PP
.B distcc
.I [COMPILER OPTIONS]
.PP
.B <compiler>
.I [COMPILER OPTIONS]
.PP
.B distcc
.I [DISTCC OPTIONS]
.SH "DESCRIPTION"
.P
distcc distributes compilation of C code across several machines on a
network.  distcc should always generate the same results as a local
compile, it is simple to install and use, and it is often much faster than a
local compile.
.PP
This version incorporates plain distcc as well as an enhancement called
pump mode or distcc-pump.
.PP
For each job, distcc in plain mode sends the complete preprocessed source code
and compiler arguments across the network from the client to a compilation
server.  In pump mode, distcc sends the source code and recursively included
header files (excluding those from the default system header directories),
so that both preprocessing and compilation can take place on the
compilation servers. This speeds up the delivery of compilations by up to an
order of magnitude over plain distcc.
.PP
Compilation is driven by a client machine, which is typically the developer's
workstation or laptop.  The distcc client runs on this machine, as does make,
the preprocessor (if distcc's pump mode is not used), the linker, and other
stages of the build process.  Any number of volunteer machines act as
compilation servers and help the client to build the program, by running the
.B distccd(1)
daemon, C compiler and assembler
as required.
.PP
distcc can run across either TCP sockets (on port 3632 by default), or
through a tunnel command such as ssh(1).  For TCP connections the
volunteers must run the distccd(1) daemon either directly or from inetd.
For SSH connections distccd must be installed but should
.B not
be listening for connections.
.PP
TCP connections should only be used on secure networks because there
is no user authentication or protection of source or object code.  SSH
connections are slower.
.PP
distcc is intended to be used with GNU Make's
.B -j
option, which runs several compiler processes concurrently.  distcc
spreads the jobs across both local and remote CPUs.  Because distcc is
able to distribute most of the work across the network, a higher
concurrency level can be used than for local builds.  As a rule of thumb,
the
.B -j
value should be set to about twice the total number of available server
CPUs but subject to client limitations.  This setting allows for maximal
interleaving of tasks being blocked waiting for disk or network IO. Note that
distcc can also work with other build control tools, such as SCons, where similar
concurrency settings must be adjusted.
.PP
The
.B -j
setting, especially for large values of
.B -j,
must take into account the CPU load
on the client.  Additional measures may be needed to curtail the client load.
For example, concurrent linking should be severely curtailed using auxiliary
locks.  The effect of other build activity, such as Java compilation when
building mixed code, should be considered.  The
.B --localslots_cpp
parameter is by default set to 8.
This limits the number of concurrent processes that do preprocessing in
plain distcc (non-pump) mode.
Therefore, larger
.B -j
values than 8 may be used without overloading a single-CPU
client due to preprocessing.  Such large values may speed up parts of the build
that do not involve C compilations, but they may not be useful to distcc
efficiency in plain mode.
.PP
In contrast, using pump mode and say 40 servers, a setting of
.B -j80
or larger may be appropriate even for single-CPU clients.
.PP
It is strongly recommended that you install the same compiler version
on all machines participating in a build.  Incompatible compilers may
cause mysterious compile or link failures.
.SH "QUICKSTART"
.TP
1
For each machine, download distcc, unpack, and install.
.TP
2
On each of the servers, run
.B distccd --daemon
with
.B --allow
options to restrict access.
.TP
3
Put the names of the servers in your environment:
.RS
$ export DISTCC_HOSTS='localhost red green blue'
.RE
.TP
4
Build!
.RS
$ make -j8 CC=distcc
.RE
.SH "QUICKSTART FOR DISTCC-PUMP MODE"
Proceed as above, but in Step 3, specify that the remote hosts are to carry the
burden of preprocessing and that the files sent over the network should be
compressed:

.RS
$ export DISTCC_HOSTS='--randomize localhost red,cpp,lzo green,cpp,lzo blue,cpp,lzo'
.RE

The
.B --randomize
option enforces a uniform usage of compile servers.  While you will
get some benefit from distcc's pump mode with only a few servers, you
get increasing benefit with more server CPUs (up to the hundreds!).
Wrap your build inside the pump command, here assuming 10 servers:

.RS
$ pump make -j20 CC=distcc
.RE

.SH "QUICKSTART FOR DISTCC-GSSAPI MODE"
Proceed as per the
.B QUICKSTART
but in Step 3, specify that the remote hosts are to mutually
authenticate with the client:

.RS
$ export DISTCC_HOSTS='--randomize localhost red,auth green,auth blue,auth'
.RE

If distccd runs under a specific principal name then execute the
following command prior to step 4:

.RS
export DISTCC_PRINICIPAL=<name>
.RE

.SH "HOW PLAIN (NON-PUMP) DISTCC WORKS"
distcc only ever runs the compiler and assembler remotely.  With plain distcc,
the preprocessor must always run locally because it needs to access various
header files on the local machine which may not be present, or may not be the
same, on the volunteer.  The linker similarly needs to examine libraries and
object files, and so must run locally.
.PP
The compiler and assembler take only a single input file (the
preprocessed source) and produce a single output (the object file).
distcc ships these two files across the network and can therefore run
the compiler/assembler remotely.
.PP
Fortunately, for most programs running the preprocessor is
relatively cheap, and the linker is called relatively
infrequent, so most of the work can be distributed.
.PP
distcc examines its command line to determine which of these
phases are being invoked, and whether the job can be
distributed.

.SH "HOW DISTCC-PUMP MODE WORKS"
In pump mode, distcc runs the preprocessor remotely too.  To do so, the
preprocessor must have access to all the files that it would have accessed if
had been running locally.  In pump mode, therefore, distcc gathers all of the
recursively included headers, except the ones that are default system headers,
and sends them along with the source file to the compilation server.
.PP
In distcc-pump mode, the server unpacks the set of all source files in a
temporary directory, which contains a directory tree that mirrors the part of
the file system that is relevant to preprocessing, including symbolic links.
.PP
The compiler is then run from the path in the temporary directory that
corresponds to the current working directory on the client.  To find and
transmit the many hundreds of files that are often part of a single compilation,
pump mode uses an incremental include analysis algorithm.  The include server is
a Python program that implements this algorithm.  The pump command starts the
include server so that throughout the build it can answer include queries by
distcc commands.
.PP
The include server uses static analysis of the macro language to deal
with conditional compilation and computed includes.  It uses the
property that when a given header file has already been analyzed for
includes, it is not necessary to do so again if all the include
options (-I's) are unchanged (along with other conditions).
.PP
For large builds, header files are included, on average, hundreds of
times each. With distcc-pump mode each such file is analyzed only a
few times, perhaps just once, instead of being preprocessed hundreds
of times.  Also, each source or header file is now compressed only
once, because the include server memoizes the compressed files.  As a
result, the time used for preparing compilations may drop by up to an
order of magnitude over the preprocessing of plain distcc.
.PP
Because distcc in pump mode is able to push out files up to about ten
times faster, build speed may increase 3X or more for large builds
compared to plain distcc mode.

.SH "RESTRICTIONS FOR PUMP MODE"

Using pump mode requires both client and servers to use release 3.0 or
later of distcc and distccd (respectively).
.PP
The incremental include analysis of distc-pump mode rests on
the fundamental assumption that source and header files do not change
during the build process.  A few complex build systems, such as that
for Linux kernel 2.6, do not quite satisfy this requirement.  To
overcome such issues, and other corner cases such as absolute
filepaths in includes, see the
.BR include_server(1)
man page.
.PP
Another important assumption is that the include configuration of all machines
must be identical.  Thus the headers under the default system path must be the
same on all servers and all clients.  If a standard GNU compiler installation
is used, then this requirement applies to all libraries whose header files are
installed under /usr/include or /usr/local/include/.  Note that installing
software packages often lead to additional headers files being placed in
subdirectories of either.
.PP
If this assumption does not hold, then it is possible to break builds with
distcc-pump mode, or worse, to get wrong results without warning.  Presently
this condition is not verified, and it is on our TODO list to address this
issue.
.PP
An easy way to guarantee that the include configurations are identical is to use
a cross-compiler that defines a default system search path restricted to
directories of the compiler installation.
.PP
See the \fBinclude_server\fR(1) manual for more information on symptoms and
causes of violations of distcc-pump mode assumptions.

.SH "HOW DISTCC-GSSAPI MODE WORKS"

In this mode distcc will use the GSS-API framework to access the currently
configured security mechanism and perform mutual authentication with the
daemon.

.SH "OPTION SUMMARY"
Most options passed to distcc are interpreted as compiler options.
The following options are understood by distcc itself.
If any of these options are specified, distcc will not invoke the
compiler.

.TP
.B --help
Displays summary instructions.
.PP
.TP
.B --version
Displays the distcc client version.
.PP
.TP
.B --show-hosts
Displays the host list that distcc would use.
See the Host Specifications section.
.PP
.TP
.B --scan-includes
Displays the list of files that distcc would send to the
remote machine, as computed by the include server.  This is a conservative
(over-)approximation of the files that would be read by the C compiler.
This option only works in pump mode.  See the "How Distcc-pump Mode Works"
section for details on how this is computed.
.RS
.P
The list output by \fBdistcc --scan-includes\fR will
contain one entry per line.  Each line contains a category followed by a path.
The category is one of FILE, SYMLINK, DIRECTORY, or SYSTEMDIR:
.IP
.B FILE
indicates a source file or header file that would
be sent to the distcc server host.
.IP
.B SYMLINK
indicates a symbolic link that would be sent to
the distcc server host.
.IP
.B DIRECTORY
indicates a directory that may be needed in order to
compile the source file.  For example, a directory "foo" may be needed
because of an include of the form #include "foo/../bar.h".
Such directories would be created on the distcc server host.
.IP
.B SYSTEMDIR
indicates a system include directory, i.e. a directory
which is on the compiler's default include path, such as "/usr/include";
such directories are assumed to be present on the distcc server host,
and so would not be sent to the distcc server host.
.RE
.TP
.B -j
Displays distcc's concurrency level, as calculated from the host list;
it is the maximum number of outstanding jobs issued by this client
to all servers.
By default this will be four times the number of hosts in the host list,
unless the /LIMIT option was used in the host list.
See the Host Specifications section.
.PP
.TP
.B --show-principal
Displays the name of the distccd security principal extracted from the
environment.
.B This option is only available if distcc was compiled with
.B the --with-auth configure option.

.SH "INSTALLING DISTCC"
There are three different ways to call distcc, to suit different
circumstances:
.RS
.PP
distcc can be installed under the name of the real compiler, to
intercept calls to it and run them remotely.  This "masqueraded"
compiler has the widest compatibility with existing source trees, and
is convenient when you want to use distcc for all compilation.  The
fact that distcc is being used is transparent to the makefiles.
.PP
distcc can be prepended to compiler command lines, such as "distcc cc
-c hello.c" or CC="distcc gcc".  This is convenient when you want to
use distcc for only some compilations or to try it out, but can cause
trouble with some makefiles or versions of libtool that assume $CC
does not contain a space.
.PP
Finally, distcc can be used directly as a compiler.  "cc" is always
used as the name of the real compiler in this "implicit" mode.  This
can be convenient for interactive use when "explicit" mode does not
work but is not really recommended for new use.
.RE
.PP
Remember that you should not use two methods for calling distcc at the
same time.  If you are using a masquerade directory, don't change CC and/or
CXX, just put the directory early on your PATH.  If you're not using
a masquerade directory, you'll need to either change CC and/or CXX, or
modify the makefile(s) to call distcc explicitly.
.SH "MASQUERADING"
The basic idea is to create a "masquerade directory" which contains
links from the name of the real compiler to the distcc binary.  This
directory is inserted early on the PATH, so that calls to the compiler
are intercepted and distcc is run instead.  distcc then removes itself
from the PATH to find the real compiler.
.PP
For example:
.PP
.RS
.nf
# mkdir /usr/lib/distcc/bin
# cd /usr/lib/distcc/bin
# ln -s ../../../bin/distcc gcc
# ln -s ../../../bin/distcc cc
# ln -s ../../../bin/distcc g++
# ln -s ../../../bin/distcc c++
.fi
.RE
.PP
Then, to use distcc, a user just needs to put the directory
/usr/lib/distcc/bin early in the PATH, and have set a host list in
DISTCC_HOSTS or a file.  distcc will handle the rest.
.PP
To automatically discover compilers and create masquerade links run
the provided
.BR update-distcc-symlinks
script.
.PP
Note that this masquerade directory must occur on the PATH earlier
than the directory that contains the actual compilers of the same
names, and that any auxiliary programs that these compilers call (such
as as or ld) must also be found on the PATH in a directory after the
masquerade directory since distcc calls out to the real compiler with
a PATH value that has all directory up to and including the masquerade
directory trimmed off.
.PP
It is possible to get a "recursion error" in masquerade mode, which
means that distcc is somehow finding itself again, not the real
compiler.  This can indicate that you have two masquerade directories
on the PATH, possibly because of having two distcc installations in
different locations.  It can also indicate that you're trying to mix
"masqueraded" and "explicit" operation.
.PP
Recursion errors can be avoided by using shell scripts instead of links.
For example, in /usr/lib/distcc/bin create a file cc which contains:
.PP
.RS
.nf
#!/bin/sh
distcc /usr/bin/gcc "$@"
.fi
.RE
.PP
In this way, we are not dependent on distcc having to locate the real gcc by
investigating the PATH variable. Instead, the compiler location is explicitly
provided.

.SH "USING DISTCC WITH CCACHE"
ccache is a program that speeds software builds by caching the results
of compilations.  ccache is normally called before distcc, so that
results are retrieved from a normal cache.  Some experimentation may
be required for idiosyncratic makefiles to make everything work together.
.PP
The most reliable method is to set
.IP
.B CCACHE_PREFIX="distcc"
.PP
This tells ccache to run distcc as a wrapper around the real
compiler.  ccache still uses the real compiler to detect compiler
upgrades.
.PP
ccache  can then be run using either a masquerade directory
.I or
by
setting
.IP
.B CC="ccache gcc"
.PP
As of version 2.2, ccache does not cache compilation from preprocessed
source and so will never get a cache hit if it is run from distccd or
distcc.  It must be run only on the client side and before distcc to
be any use.

distcc's pump mode is not compatible with ccache.
.SH "HOST SPECIFICATIONS"
A "host list" tells distcc which machines to use for compilation.  In
order, distcc looks in the
.B $DISTCC_HOSTS
environment variable, the user's
.B $DISTCC_DIR/hosts
file, and the system-wide host
file.  If no host list can be found, distcc emits a warning and
compiles locally.
.PP
The host list is a simple whitespace separated list of host
specifications.  The simplest and most common form is a host names,
such as
.PP
.RS
.B localhost red green blue
.RE
.PP
distcc prefers hosts towards the start of the list, so machines should
be listed in descending order of speed.  In particular, when only a
single compilation can be run (such as from a configure script), the
first machine listed is used (but see
.I --randomize
below).
.PP
Placing
.I localhost
at the right point in the list is important to getting good
performance.  Because overhead for running jobs locally is low,
localhost should normally be first.  However, it is important that the
client have enough cycles free to run the local jobs and the distcc
client.  If the client is slower than the volunteers, or if there are
many volunteers, then the client should be put later in the list or
not at all.  As a general rule, if the aggregate CPU speed of the
client is less than one fifth of the total, then the client should be
left out of the list.
.PP
If you have a large shared build cluster and a single shared hosts file,
the above rules would cause the first few machines in the hosts
file to be tried first even though they are likely to be busier than machines
later in the list.  To avoid this, place the keyword
.I --randomize
into the host list.  This will cause the host list to be randomized,
which should improve performance slightly for large build clusters.
.PP
There are two special host names
.B --localslots
and
.B --localslots_cpp
which are useful for adjusting load on the local machine.  The
.B --localslots
host specifies how many jobs that cannot be run remotely that can be run concurrently
on the local machine, while
.B --localslots_cpp
controls how many preprocessors will run in parallel on the local machine.  Tuning
these values can improve performance.  Linking on large projects
can take large amounts of memory.  Running parallel linkers, which cannot be
executed remotely,  may
force the machine to swap, which reduces performance over just running the
jobs in sequence without swapping.   Getting the number of parallel preprocessors just
right allows you to use larger parallel factors with make, since the local machine
now has some mechanism for measuring local resource usage.
.PP
Finally there is the host entry
.PP
Performance depends on the details of the source and makefiles used
for the project, and the machine and network speeds.  Experimenting
with different settings for the host list and
.B -j
factor may improve performance.
.PP
The syntax is
.PP
.nf
  DISTCC_HOSTS = HOSTSPEC ...
  HOSTSPEC = LOCAL_HOST | SSH_HOST | TCP_HOST | OLDSTYLE_TCP_HOST
                        | GLOBAL_OPTION
                        | ZEROCONF
  LOCAL_HOST = localhost[/LIMIT]
             | --localslots=<int>

NEWS
distcc-3.5 "Not dead yet" 2025-11-??

  SECURITY:

    * Since v3.3, distccd has defaulted to open access despite warning it chose
      --allow-private when no --allow* given (#451, Dominique Martinet)

  CHANGES:

   * The popt library is no longer shipped: it must be separately installed
     as a dependency before building. (Martin Pool)

   * The libiberty library is no longer a dependency; providing arguments via
     @cmd_file syntax is no longer supported. (Martin Pool)

   * The pump include_server now uses setuptools instead of distutils.
     (#493, Alexey Sheplyakov)

  FEATURES:

   * (Linux) Added --oom-score-adj option to distccd to adjust the OOM Killer's
     preferences in an out-of-memory scenario. (#439, Matt Whitlock)

   * (Linux) Added dcc_free_mem reporting of available system memory in the
     distccd "stats" output. (#523, Whisperity)

   * Add `sysroot` option and allow the `-specs` option. (#531, YunQiang Su).

   * Added a `--timeout` option to include_server to configure the total cpu
     time it can take on a single closure calculation. (#559, Alex Karle)

   * Added a `--additional_system_dir` option to include_server to mark
     arbitrary directories as system directories (assumed present on all
     machines). (#559, Alex Karle)

   * Added a `--skip_parsing_system_files` option to include_server to avoid
     parsing headers in system directories (a significant speedup, provided they
     don't define macros used in include lines). (#559, Alex Karle)

   * Added a `-D` option to include_server to define additional macros
     (for macros needed from skipped system directories). (#559, Alex Karle)

   * Added a `--unsafe_no_unexpanded_macro_fns` option to include_server to
     reduce complex macro possible expansion permutations.
     (#551, Alastair Rankine and Alex Karle)

  ENHANCEMENTS

   * Files with .incbin directives compiled locally (#461, Alexey Sheplyakov)
   * Optimized include_server's directory forcing_technique (#567, Alex Karle)
   * Increased include_server's connection limit (#556, Alex Karle)
   * Support --target=arch clang cross-compilation (#417, Alexey Sheplyakov)
   * distcc --jobs argument no longer capped at 200 (#458, @00gh)
   * distccd properly handles -f*-prefix-map switches (#459, Rosen Matev)
   * Masquerade symlinks now work on rpm-based systems (#436, Alexey Sheplyakov)

  BUG FIXES:

   * Fixed --allow-private when compiled without IPv6 (#451, Dominique Martinet)
   * Fixed IPv4-based access control when listening on IPv6 (#444, Alexey Sheplyakov)
   * Reduced RAM consumed by include server for large codebases (#566, Alex Karle)
   * Fixed include_server mangling unicode characters in source (#549, Alex Karle)
   * Fixed double free crash when distcc downgrades from pump (#552, Alex Karle)
   * Fixed include_server --verify mode (#553, Alex Karle)
   * Fixed distccd crash in signal handlers seen on macOS (#564, Alex Karle)
   * Fixed distccd comp_dir path rewrite under DWARF 5 (#547, Alex Karle)
   * Fixed include_server handling of empty macros (#557, Alex Karle)
   * Fixed various bugs in the include_server macro parser (#551, Alastair Rankine)
   * Set cap on max retries with DISTCC_BACKOFF=0 to prevent inf. loop (#438, Alexey Sheplyakov)
   * Fixed cross-rewriting on non-x86 systems (#427, Alexey Sheplyakov)
   * Fixed segfault in distcc when compiler not found (#511, GeorgeFrancisco, Martin Pool)
   * Fixed heap corruption with gcc cross compilation (#425, Alexey Sheplyakov)

distcc-3.4 "Lax lexer" 2021-4-11

  FEATURES:

   * distccmon-gnome ported from gtk2 to gtk3

   * Remove debug lag in spawning new threads. (better multi-core performance)

   * Bug fixes.

distcc-3.3 "Charlie the unicorn" 2018-3-3

  FEATURES:

   * Use masquerade as compiler white-list. Bug #155 d849ca2. (Shawn Landden)

   * New --allow-private (the default) which allows non-global
     IP and IPv6 addresses. (Shawn Landden)

   * Cross-compilation support. PR #244. (Shawn Landden)

   * Fix parsing of IPv6 addresses. PR #175 (Bing Sun)

   * Python 3, not python 2.

   * Can build without python (and without pump mode or tests).

distcc-3.2 "Back in black"  2011-10-11

  FEATURES:

   * IPv6 support (adshea, Benjamin R. Haskell, Bob Ham).

   * Optional support for GSS-API authentication (Ian Baker at CERN)

   * Added an environment variable DISTCC_SKIP_LOCAL_RETRY for skipping
     the local retry in case of a remote compilation failure (Ryan Burns).

   * Some constants that were previously hard-coded in the
     sources are now configurable via environment variables:
     DISTCC_MAX_DISCREPANCY, DISTCC_IO_TIMEOUT (Lei Zhang).

  BUG FIXES:

   * Fixed the following 20 issues <http://code.google.com/p/distcc/issues/list>:

     30: 3.0 on OSX: ValueError: Expected absolute path, but got '(framework'.
     33: distcc --scan-includes dumps core
     34: Zeroconf fails if avahi supports IPv6
     35: Include server internal error: '<class 'sre_constants.error'>: ('bogus escape (end of line)',)'
     36: distcc-mon-gnome displays multiple rows for same host/slot
     39: build with gcc 4.1 snprintf.h error "expected declaration specifiers"
     42: distcc --show-hosts fails when using Avahi with IPv6 support and +zeroconf for distcc.
     43: Remove duplicate hosts from zeroconf list
     44: Failed tests on make check (gentoo gcc-4.3.3)
     46: Add TAGS target to Makefile.in
     49: make check fails on PreprocessAsm_Case
     57: Fix warnings
     58: PUMP mode can`t PARSE the HOSTS
     60: CPlusPlus test case fails with GCC 4.4.1
     61: Gdb tests fail in pumped mode on Ubuntu 9.10 amd64
     66: pump mode fails to compile certain KDE-4 packages, finds/looks for incorrect headers
     70: variable overlapping causing errors..
     75: include_server documentation, distcc_fallback flag incorrect
     81: distcc 3.1 will not compile correctly with current version of gcc

distcc-3.1 "Humming along"  2008-11-21

  FEATURES:


src/distcc.c
/* -*- c-file-style: "java"; indent-tabs-mode: nil; tab-width: 4; fill-column: 78 -*-
 *
 * distcc -- A simple distributed compiler system
 *
 * Copyright (C) 2002, 2003, 2004 by Martin Pool <mbp@samba.org>
 * Copyright 2007 Google Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 */


            /* 4: The noise of a multitude in the
             * mountains, like as of a great people; a
             * tumultuous noise of the kingdoms of nations
             * gathered together: the LORD of hosts
             * mustereth the host of the battle.
             *        -- Isaiah 13 */



#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>

#include "distcc.h"
#include "trace.h"
#include "exitcode.h"
#include "util.h"
#include "hosts.h"
#include "bulk.h"
#include "implicit.h"
#include "compile.h"
#include "emaillog.h"


/* Name of this program, for trace.c */
const char *rs_program_name = "distcc";

/**
 * @file
 *
 * Entry point for the distcc client.
 *
 * There are three methods of use for distcc: explicit (distcc gcc -c
 * foo.c), implicit (distcc -c foo.c) and masqueraded (gcc -c foo.c,
 * where gcc is really a link to distcc).
 *
 * Detecting these is relatively easy by examining the first one or
 * two words of the command.  We also need to make sure that when we
 * go to run the compiler, we run the one intended by the user.
 *
 * In particular, for masqueraded mode, we want to make sure that we
 * don't invoke distcc recursively.
 **/

static void dcc_show_usage(void)
{
    dcc_show_version("distcc");
    printf(
"Usage:\n"
"   distcc [--scan-includes] [COMPILER] [compile options] -o OBJECT -c SOURCE\n"
"   distcc [--help|--version|--show-hosts|-j]\n"
"\n"
"Options:\n"
"   COMPILER                   Defaults to \"cc\".\n"
"   --help                     Explain usage, and exit.\n"
"   --version                  Show version, and exit.\n"
"   --show-hosts               Show host list, and exit.\n"
"   -j                         Show the concurrency level, as calculated from\n"
"                              the host list, and exit.\n"
"   --scan-includes            Show the files that distcc would send to the\n"
"                              remote machine, and exit.  (Pump mode only.)\n"
#ifdef HAVE_GSSAPI
"   --show-principal           Show current distccd GSS-API principal and exit.\n"
#endif
"\n"
"Environment variables:\n"
"   See the manual page for a complete list.\n"
"   DISTCC_VERBOSE=1           Give debug messages.\n"
"   DISTCC_LOG                 Send messages to file, not stderr.\n"
"   DISTCC_SSH                 Command to run to open SSH connections.\n"
"   DISTCC_DIR                 Directory for host list and locks.\n"
#ifdef HAVE_GSSAPI
"   DISTCC_PRINCIPAL	      The name of the server principal to connect to.\n"
#endif
"\n"
"Server specification:\n"
"A list of servers is taken from the environment variable $DISTCC_HOSTS, or\n"
"$DISTCC_DIR/hosts, or ~/.distcc/hosts, or %s/distcc/hosts.\n"
"Each host can be given in any of these forms, see the manual for details:\n"
"\n"
"   localhost                  Run in place.\n"
"   HOST                       TCP connection, port %d.\n"
"   HOST:PORT                  TCP connection, specified port.\n"
"   @HOST                      SSH connection to specified host.\n"
"   USER@HOST                  SSH connection to specified username at host.\n"
"   HOSTSPEC,lzo               Enable compression.\n"
"   HOSTSPEC,cpp,lzo           Use pump mode (remote preprocessing).\n"
"   HOSTSPEC,auth              Enable GSS-API based mutual authenticaton.\n"
"   --randomize                Randomize the server list before execution.\n"
"\n"
"distcc distributes compilation jobs across volunteer machines running\n"
"distccd.  Jobs that cannot be distributed, such as linking, are run locally.\n"
"distcc should be used with make's -jN option to execute in parallel on\n"
"several machines.\n",
    SYSCONFDIR,
    DISTCC_DEFAULT_PORT);
}


static void dcc_client_signalled (int whichsig)
{
    signal(whichsig, SIG_DFL);

    dcc_log_signal_termination(whichsig);

    dcc_cleanup_tempfiles_from_signal_handler();

    raise(whichsig);

}


static void dcc_client_catch_signals(void)
{
    signal(SIGTERM, &dcc_client_signalled);
    signal(SIGINT, &dcc_client_signalled);
    signal(SIGHUP, &dcc_client_signalled);
}

static void dcc_free_hostlist(struct dcc_hostdef *list) {
    while (list) {
        struct dcc_hostdef *l = list;
        list = list->next;
        dcc_free_hostdef(l);
    }
}

static void dcc_show_hosts(void) {
    struct dcc_hostdef *list, *l;
    int nhosts;

    if (dcc_get_hostlist(&list, &nhosts) != 0) {
        rs_log_crit("Failed to get host list");
        return;
    }

    for (l = list; l; l = l->next)
        printf("%s\n", l->hostdef_string);

    dcc_free_hostlist(list);
}

static void dcc_concurrency_level(void) {
    struct dcc_hostdef *list, *l;
    int nhosts;
    int nslots = 0;

    if (dcc_get_hostlist(&list, &nhosts) != 0) {
        rs_log_crit("Failed to get host list");
        return;
    }

    for (l = list; l; l = l->next)
        nslots += l->n_slots;

    dcc_free_hostlist(list);

    printf("%i\n", nslots);
}

#ifdef HAVE_GSSAPI
/*
 * Print out the name of the principal.
 */
static void dcc_gssapi_show_principal(void) {
    char *princ_env_val = NULL;

    if((princ_env_val = getenv("DISTCC_PRINCIPAL"))) {
	    printf("Principal is\t: %s\n", princ_env_val);
    } else {
        printf("Principal\t: Not Set.\n");
    }
}
#endif

/**
 * distcc client entry point.
 *
 * This is typically called by make in place of the real compiler.
 *
 * Performs basic setup and checks for distcc arguments, and then kicks off
 * dcc_build_somewhere().
 **/
int main(int argc, char **argv)
{
    int status, sg_level, tweaked_path = 0;
    char **compiler_args = NULL; /* dynamically allocated */
    char *compiler_name; /* points into argv[0] */
    int ret;

    dcc_client_catch_signals();
    atexit(dcc_cleanup_tempfiles);
    atexit(dcc_remove_state_file);

    dcc_set_trace_from_env();
    dcc_setup_log_email();

    dcc_trace_version();

    /* TODO: Expand @file arguments? Previously this was done only when built with libiberty. */

    compiler_name = (char *) dcc_find_basename(argv[0]);

    /* Ignore SIGPIPE; we consistently check error codes and will
     * see the EPIPE. */
    dcc_ignore_sigpipe(1);

    sg_level = dcc_recursion_safeguard();

    rs_trace("compiler name is \"%s\"", compiler_name);

    if (strstr(compiler_name, "distcc") != NULL) {
        /* Either "distcc -c hello.c" or "distcc gcc -c hello.c" */
        if (argc <= 1) {
            fprintf (stderr,
                     "%s: missing option/operand\n"
                     "Try `%s --help' for more information.\n",
                     argv[0], argv[0]);
            ret = EXIT_BAD_ARGUMENTS;
            goto out;
        }

        if (!strcmp(argv[1], "--help")) {
            dcc_show_usage();
            ret = 0;
            goto out;
        }

        if (!strcmp(argv[1], "--version")) {
            dcc_show_version("distcc");
            ret = 0;
            goto out;
        }

        if (!strcmp(argv[1], "--show-hosts")) {
            dcc_show_hosts();
            ret = 0;
            goto out;
        }

        if (!strcmp(argv[1], "-j")) {
            dcc_concurrency_level();
            ret = 0;
            goto out;
        }

        if (!strcmp(argv[1], "--scan-includes")) {
            if (argc <= 2) {
                fprintf (stderr,
                         "%s: missing operand\n"
                         "Try `%s --help' for more information.\n",
                         argv[0], argv[0]);
                ret = EXIT_BAD_ARGUMENTS;
                goto out;
            }
            dcc_scan_includes = 1;
            argv++;
        }

#ifdef HAVE_GSSAPI
	    if (!strcmp(argv[1], "--show-principal")) {
	        dcc_gssapi_show_principal();
	        ret = 0;
	        goto out;
	    }
#endif

        if ((ret = dcc_find_compiler(argv, &compiler_args)) != 0) {
            goto out;
        }
        /* compiler_args is now respectively either "cc -c hello.c" or
         * "gcc -c hello.c" */

#if 0
        /* I don't think we need to call this: if we reached this
         * line, our invocation name is something like 'distcc', and
         * that's never a problem for masquerading loops. */
        if ((ret = dcc_trim_path(compiler_name)) != 0)
            goto out;
#endif
    } else {
        /* Invoked as "cc -c hello.c", with masqueraded path */
        if ((ret = dcc_support_masquerade(argv, compiler_name,
                                          &tweaked_path)) != 0)
            goto out;

        if ((ret = dcc_copy_argv(argv, &compiler_args, 0)) != 0) {
            goto out;
        }
        free(compiler_args[0]);
        compiler_args[0] = strdup(compiler_name);
        if (!compiler_args[0]) {
            rs_log_error("strdup failed - out of memory?");
            ret = EXIT_OUT_OF_MEMORY;
            goto out;
        }
    }

    if (sg_level - tweaked_path > 0) {
        rs_log_crit("distcc seems to have invoked itself recursively!");
        ret = EXIT_RECURSION;
        goto out;
    }

    ret = dcc_build_somewhere_timed(compiler_args, sg_level, &status);
    compiler_args = NULL; /* dcc_build_somewhere_timed already free'd it. */

    out:
    if (compiler_args) {
      dcc_free_argv(compiler_args);
    }
    dcc_maybe_send_email();
    dcc_exit(ret);
}
